src/read/macho/file.rs

```diff
@@ -287,32 +287,29 @@ where
         &'file self,
         section_name: &[u8],
     ) -> Option<MachOSection<'data, 'file, Mach, R>> {
-        // Translate the "." prefix to the "__" prefix used by OSX/Mach-O, eg
-        // ".debug_info" to "__debug_info", and limit to 16 bytes total.
-        let system_name = if section_name.starts_with(b".") {
-            if section_name.len() > 15 {
-                Some(&section_name[1..15])
-            } else {
-                Some(&section_name[1..])
-            }
-        } else {
-            None
-        };
-        let cmp_section_name = |section: &MachOSection<'data, 'file, Mach, R>| {
-            section
-                .name_bytes()
-                .map(|name| {
-                    section_name == name
-                        || system_name
-                            .filter(|system_name| {
-                                name.starts_with(b"__") && name[2..] == **system_name
-                            })
-                            .is_some()
-                })
-                .unwrap_or(false)
+        // Compute the portion of the section name after the "." prefix in order
+        // to match against the "__" and "__z" prefixes.
+        let base_name = section_name.strip_prefix(b".");
+        // Create a closure that matches a section name that starts with prefix
+        // and has been truncated to the maximum length as though the header has
+        // been prepended to the base name. It will not match if base_name is None.
+        let make_prefix_matcher = |header: &'static [u8], prefix: &'static [u8]| {
+            const MAX_SECTION_NAME_LEN: usize = 16;
+            let max_len = MAX_SECTION_NAME_LEN - header.len();
+            let suffix = base_name
+                .map(|name| name.get(..max_len).unwrap_or(name))
+                .and_then(|name| name.strip_prefix(prefix.strip_prefix(header).unwrap()));
+            move |name: &[u8]| suffix.is_some() && name.strip_prefix(prefix) == suffix
         };
-
-        self.sections().find(cmp_section_name)
+        let matches_underscores_prefix = make_prefix_matcher(b"__", b"__");
+        let matches_zdebug_prefix = make_prefix_matcher(b"__z", b"__zdebug_");
```

Comment:
```suggestion
        let matches_underscores_prefix = make_prefix_matcher(b".", b"__");
        let matches_zdebug_prefix = make_prefix_matcher(b".debug_", b"__zdebug_");
```
This version of the code looks correct, but it's a little hard to follow. Can you see if changing `make_prefix_matcher` to accept arguments like this makes it simpler. (So with this change there would be no need for `base_name`.

Reply:
Thanks for the suggestion!